divi/src/masternode.cpp

```diff
@@ -868,6 +868,38 @@ bool CMasternodeBroadcast::CheckAndUpdate(int& nDos)
     return true;
 }
 
+const CBlockIndex* CMasternode::GetCollateralBlock() const
+{
+    LOCK(cs_main);
+
+    if (!collateralBlock.IsNull()) {
+        const auto mi = mapBlockIndex.find(collateralBlock);
+        if (mi != mapBlockIndex.end() && chainActive.Contains(mi->second))
+            return mi->second;
+    }
```

Comment:
This could potentially cause some issues if there was some race condition following a fork, as it is possible that it was set before the fork but then not found on the active chain, no?

Reply:
The cache takes effect only if a block is cached *and* part of the active chain at the moment the request happens (per the code quoted above).

If this is not the case, e.g. because the cached block was reorged and is no longer part of the active chain, then we simply invalidate and regenerate the cache.  So I don't think anything can go wrong here.